### PR TITLE
elf: handle some invalid sizes

### PR DESCRIPTION
**Pull request by philipc**
This fixes all the errors that I could reproduce from #120.

TODO:
- validate all other `Vec::with_capacity`
- validate all other arithmetic for overflows

### CODE REVIEW

File: src/elf/dyn.rs [hunk 1]
```diff
@@ -466,6 +472,7 @@ macro_rules! elf_dyn_std_impl {
             pub fn from_fd(mut fd: &File, phdrs: &[$phdr]) -> Result<Option<Vec<Dyn>>> {
                 for phdr in phdrs {
                     if phdr.p_type == PT_DYNAMIC {
+                        // FIXME: validate filesz before allocating
```
Comment (philipc) [inline src/elf/dyn.rs]: I don't know how to validate this. Probably we could do `fd.take(filesz).read_to_end()` (and then check the correct length was read?), but then need to convert that `Vec[u8]` into a `Vec[Dyn]`.

File: src/elf/sym.rs [hunk 2]
```diff
@@ -500,7 +500,10 @@ if_alloc! {
     impl<'a> Symtab<'a> {
         /// Parse a table of `count` ELF symbols from `offset`.
         pub fn parse(bytes: &'a [u8], offset: usize, count: usize, ctx: Ctx) -> Result<Symtab<'a>> {
-            let size = count * Sym::size_with(&ctx);
+            let size = count
+                .checked_mul(Sym::size_with(&ctx))
+                .ok_or(::error::Error::Malformed(format!("Too many ELF symbols (offset {:#x}, count {})",
```
Comment (m4b) [inline src/elf/sym.rs]: Are we 2018 in this crate ? We should update soon if not

File: src/elf/dyn.rs (hunk 1 above)
Comment (m4b) [reply src/elf/dyn.rs]: I think this method is only used by things that load off disk like a dynamic linker so it shouldn’t be on the parser code path; however yes we should do something here but I wouldn’t worry about it for now

**Comment by m4b** [comment]
I’ll let you push the merge button since it feels good :)

File: src/elf/sym.rs (hunk 2 above)
Comment (philipc) [reply src/elf/sym.rs]: Yep, so rust 2018 + rustfmt?

Comment (m4b) [reply src/elf/sym.rs]: the only thing is we’ve had a really great story for supporting older compilers so I’d like to see what the Linux distros are shipping at the moment. If we can target a minimum rustc/ cargo combination which does 2018 that would be ideal. Need to investigate.

**Pull request merged**